# **PROJECT NO 3**

**Author Name:** *Ahsanullah Ghani*

**Email:** *ahsanullahghani000@gmail.com*

 **The data was Collected from the following:** [Link](https://www.kaggle.com/datasets/yasserh/nyc-taxi-trip-duration)

## Problem Statement

New York City is one of the busiest cities in the world, and thousands of taxi trips take place every day. The time required to complete a taxi trip can vary greatly, even when the distance between two locations is almost the same. Traffic, time of day, day of the week, pickup and drop-off locations, and other factors can all affect how long a trip takes.

In this project, I am working with a large dataset of NYC taxi trips to understand the factors that influence trip duration. Instead of focusing only on building a highly complex prediction model, the main goal is to explore the data, identify useful patterns, and create meaningful features from the available information.

By analyzing factors such as trip distance, pickup and drop-off locations, time, day, and travel direction, this project aims to understand why some taxi trips take longer than others. The analysis can help provide useful insights into taxi travel patterns and the factors that have the greatest impact on trip duration.

The overall goal of this project is to turn raw taxi trip data into meaningful information that can help us better understand and predict travel time in New York City.


<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **TASK**

</div>

<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **1: DATA LOADING**

</div>

***Import Library***

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px


***Import Dataset***

In [2]:
df = pd.read_csv(r"C:\Users\AHSAN GHANI\Documents\6Month Intership program\Frist Month\pandas_tip&trick\Data set\Taxi trip duration\NYC.csv")

***Check the First 5 Row***

In [ ]:
df.head(100)

,id,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration
0,id2875421,2,2016-03-14 17:24:55,2016-03-14 17:32:30,1,-73.982155,40.767937,-73.964630,40.765602,N,455
1,id2377394,1,2016-06-12 00:43:35,2016-06-12 00:54:38,1,-73.980415,40.738564,-73.999481,40.731152,N,663
2,id3858529,2,2016-01-19 11:35:24,2016-01-19 12:10:48,1,-73.979027,40.763939,-74.005333,40.710087,N,2124
3,id3504673,2,2016-04-06 19:32:31,2016-04-06 19:39:40,1,-74.010040,40.719971,-74.012268,40.706718,N,429
4,id2181028,2,2016-03-26 13:30:55,2016-03-26 13:38:10,1,-73.973053,40.793209,-73.972923,40.782520,N,435


***Check Dataset Shape & Columns***

In [5]:
df.shape

(1458644, 11)

In [6]:
df.columns

Index(['id', 'vendor_id', 'pickup_datetime', 'dropoff_datetime',
       'passenger_count', 'pickup_longitude', 'pickup_latitude',
       'dropoff_longitude', 'dropoff_latitude', 'store_and_fwd_flag',
       'trip_duration'],
      dtype='str')

In [10]:
df.dtypes

id                        str
vendor_id               int64
pickup_datetime           str
dropoff_datetime          str
passenger_count         int64
pickup_longitude      float64
pickup_latitude       float64
dropoff_longitude     float64
dropoff_latitude      float64
store_and_fwd_flag        str
trip_duration           int64
dtype: object

<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **2: DATA CLEANING**

</div>

***Check missing values using .isnull().sum()***

In [7]:
df.isnull().sum()

id                    0
vendor_id             0
pickup_datetime       0
dropoff_datetime      0
passenger_count       0
pickup_longitude      0
pickup_latitude       0
dropoff_longitude     0
dropoff_latitude      0
store_and_fwd_flag    0
trip_duration         0
dtype: int64

***Handle missing values***

In [8]:
df.isnull().mean() * 100

id                    0.0
vendor_id             0.0
pickup_datetime       0.0
dropoff_datetime      0.0
passenger_count       0.0
pickup_longitude      0.0
pickup_latitude       0.0
dropoff_longitude     0.0
dropoff_latitude      0.0
store_and_fwd_flag    0.0
trip_duration         0.0
dtype: float64

In [8]:
df['passenger_count'].fillna(df['passenger_count'].mean(),inplace=True)

C:\Users\AHSAN GHANI\AppData\Local\Temp\ipykernel_23020\3567702229.py:1: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df['passenger_count'].fillna(df['passenger_count'].mean(),inplace=True)


0          1
1          1
2          1
3          1
4          1
          ..
1458639    4
1458640    1
1458641    1
1458642    1
1458643    1
Name: passenger_count, Length: 1458644, dtype: int64

In [12]:
df.isnull().mean() * 100

id                    0.0
vendor_id             0.0
pickup_datetime       0.0
dropoff_datetime      0.0
passenger_count       0.0
pickup_longitude      0.0
pickup_latitude       0.0
dropoff_longitude     0.0
dropoff_latitude      0.0
store_and_fwd_flag    0.0
trip_duration         0.0
dtype: float64

In [17]:
df.dropna(subset=['passenger_count'], inplace=True)

<div style="
    width: 92%;
    max-width: 1200px;
    margin: 18px auto;
    padding: 22px 26px;
    background: #101820;
    border: 2px solid #00aaff;
    border-radius: 14px;
    box-sizing: border-box;
    box-shadow:
        0 0 8px #00aaff,
        0 0 18px rgba(0, 170, 255, 0.8),
        0 0 30px rgba(0, 170, 255, 0.4),
        inset 0 0 14px rgba(0, 170, 255, 0.15);
">

<h3 style="
    color: #00bfff;
    margin: 0 0 12px 0;
    font-size: clamp(20px, 2.5vw, 28px);
    text-shadow:
        0 0 6px #00aaff,
        0 0 12px #0088cc;
">
Observation
</h3>

<p style="
    color: #f2f2f2;
    font-size: clamp(14px, 1.3vw, 17px);
    line-height: 1.8;
    margin: 0;
">

I checked the dataset for missing values to understand whether any data was missing.
The dataset did not contain any missing values, so no missing-value treatment was required.

However, I practiced different techniques such as

<code style="
    color: #00c8ff;
    background: #0b2735;
    padding: 4px 7px;
    border-radius: 5px;
    font-weight: bold;
    text-shadow: 0 0 5px #00aaff;
">
dropna()
</code>

and

<code style="
    color: #00c8ff;
    background: #0b2735;
    padding: 4px 7px;
    border-radius: 5px;
    font-weight: bold;
    text-shadow: 0 0 5px #00aaff;
">
fillna()
</code>

to review and strengthen my understanding of how missing values are handled in data analysis.

</p>

</div>

***Handle Duplicated***

In [20]:
df.duplicated().sum()

np.int64(0)

In [ ]:
df[df.duplicated()]# Check the duplicated row

,id,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration


In [22]:
df.drop_duplicates(inplace=True)

In [23]:
df.duplicated().sum()

np.int64(0)

<div style="
    width: 92%;
    max-width: 1200px;
    margin: 18px auto;
    border: 2px solid #00aaff;
    border-radius: 14px;
    padding: 22px 26px;
    background: #101820;
    box-shadow:
        0 0 8px #00aaff,
        0 0 18px rgba(0, 170, 255, 0.8),
        0 0 30px rgba(0, 170, 255, 0.4),
        inset 0 0 14px rgba(0, 170, 255, 0.15);
    box-sizing: border-box;
">

<h3 style="
    color: #00bfff;
    margin-top: 0;
    margin-bottom: 12px;
    font-size: clamp(20px, 2.5vw, 28px);
    text-shadow:
        0 0 6px #00aaff,
        0 0 12px #0088cc;
">
Observation
</h3>

<p style="
    color: #f2f2f2;
    font-size: clamp(14px, 1.3vw, 17px);
    line-height: 1.8;
    margin-bottom: 0;
">

I checked the dataset for duplicate records using
<code style="
    color:#00c8ff;
    background:#0b2735;
    padding:3px 6px;
    border-radius:4px;
">
duplicated().sum()
</code>.

The result was
<strong style="
    color:#00bfff;
    text-shadow:0 0 6px #00aaff;
">
0
</strong>,
which means there were no completely duplicated rows in the dataset.
Therefore, no duplicate records needed to be removed.

</p>

</div>

***- Convert Order Date and Ship Date into datetime*** 

In [25]:
##Check order Data and Ship Date
df.head()

,id,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration
0,id2875421,2,2016-03-14 17:24:55,2016-03-14 17:32:30,1,-73.982155,40.767937,-73.964630,40.765602,N,455
1,id2377394,1,2016-06-12 00:43:35,2016-06-12 00:54:38,1,-73.980415,40.738564,-73.999481,40.731152,N,663
2,id3858529,2,2016-01-19 11:35:24,2016-01-19 12:10:48,1,-73.979027,40.763939,-74.005333,40.710087,N,2124
3,id3504673,2,2016-04-06 19:32:31,2016-04-06 19:39:40,1,-74.010040,40.719971,-74.012268,40.706718,N,429
4,id2181028,2,2016-03-26 13:30:55,2016-03-26 13:38:10,1,-73.973053,40.793209,-73.972923,40.782520,N,435


In [26]:
df[['pickup_datetime','dropoff_datetime']]

,pickup_datetime,dropoff_datetime
0,2016-03-14 17:24:55,2016-03-14 17:32:30
1,2016-06-12 00:43:35,2016-06-12 00:54:38
2,2016-01-19 11:35:24,2016-01-19 12:10:48
3,2016-04-06 19:32:31,2016-04-06 19:39:40
4,2016-03-26 13:30:55,2016-03-26 13:38:10
...,...,...
1458639,2016-04-08 13:31:04,2016-04-08 13:44:02
1458640,2016-01-10 07:35:15,2016-01-10 07:46:10
1458641,2016-04-22 06:57:41,2016-04-22 07:10:25
1458642,2016-01-05 15:56:26,2016-01-05 16:02:39


In [27]:
# Now apply datetime
df[['pickup_datetime','dropoff_datetime']] =df[['pickup_datetime','dropoff_datetime']].apply(pd.to_datetime)

In [28]:
df[['pickup_datetime','dropoff_datetime']]

,pickup_datetime,dropoff_datetime
0,2016-03-14 17:24:55,2016-03-14 17:32:30
1,2016-06-12 00:43:35,2016-06-12 00:54:38
2,2016-01-19 11:35:24,2016-01-19 12:10:48
3,2016-04-06 19:32:31,2016-04-06 19:39:40
4,2016-03-26 13:30:55,2016-03-26 13:38:10
...,...,...
1458639,2016-04-08 13:31:04,2016-04-08 13:44:02
1458640,2016-01-10 07:35:15,2016-01-10 07:46:10
1458641,2016-04-22 06:57:41,2016-04-22 07:10:25
1458642,2016-01-05 15:56:26,2016-01-05 16:02:39


In [29]:
df[['pickup_datetime','dropoff_datetime']].dtypes

pickup_datetime     datetime64[us]
dropoff_datetime    datetime64[us]
dtype: object

 ***Remove invalid rows (e.g. zero distance, zero duration)*** 

In [34]:
df = df[
    (df["trip_duration"] > 0) &
    (
        (df["pickup_longitude"] != df["dropoff_longitude"]) |
        (df["pickup_latitude"] != df["dropoff_latitude"])
    )
]
df

,id,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration
0,id2875421,2,2016-03-14 17:24:55,2016-03-14 17:32:30,1,-73.982155,40.767937,-73.964630,40.765602,N,455
1,id2377394,1,2016-06-12 00:43:35,2016-06-12 00:54:38,1,-73.980415,40.738564,-73.999481,40.731152,N,663
2,id3858529,2,2016-01-19 11:35:24,2016-01-19 12:10:48,1,-73.979027,40.763939,-74.005333,40.710087,N,2124
3,id3504673,2,2016-04-06 19:32:31,2016-04-06 19:39:40,1,-74.010040,40.719971,-74.012268,40.706718,N,429
4,id2181028,2,2016-03-26 13:30:55,2016-03-26 13:38:10,1,-73.973053,40.793209,-73.972923,40.782520,N,435
...,...,...,...,...,...,...,...,...,...,...,...
1458639,id2376096,2,2016-04-08 13:31:04,2016-04-08 13:44:02,4,-73.982201,40.745522,-73.994911,40.740170,N,778
1458640,id1049543,1,2016-01-10 07:35:15,2016-01-10 07:46:10,1,-74.000946,40.747379,-73.970184,40.796547,N,655
1458641,id2304944,2,2016-04-22 06:57:41,2016-04-22 07:10:25,1,-73.959129,40.768799,-74.004433,40.707371,N,764
1458642,id2714485,1,2016-01-05 15:56:26,2016-01-05 16:02:39,1,-73.982079,40.749062,-73.974632,40.757107,N,373


<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **4: FEATURE SCALING / NORMALIZATION**

</div>

***Apply Min-Max scaling***

In [35]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

df[["trip_duration", "passenger_count"]] = scaler.fit_transform(
    df[["trip_duration", "passenger_count"]]
)

In [36]:
df.columns

Index(['id', 'vendor_id', 'pickup_datetime', 'dropoff_datetime',
       'passenger_count', 'pickup_longitude', 'pickup_latitude',
       'dropoff_longitude', 'dropoff_latitude', 'store_and_fwd_flag',
       'trip_duration'],
      dtype='str')

In [37]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
df[['pickup_longitude','pickup_latitude']] = scaler.fit_transform(df[['pickup_longitude','pickup_latitude']])


In [39]:
df.head(10)

,id,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration
0,id2875421,2,2016-03-14 17:24:55,2016-03-14 17:32:30,0.125,0.872456,0.352714,-73.964630,40.765602,N,0.000129
1,id2377394,1,2016-06-12 00:43:35,2016-06-12 00:54:38,0.125,0.872487,0.351004,-73.999481,40.731152,N,0.000188
2,id3858529,2,2016-01-19 11:35:24,2016-01-19 12:10:48,0.125,0.872512,0.352482,-74.005333,40.710087,N,0.000602
3,id3504673,2,2016-04-06 19:32:31,2016-04-06 19:39:40,0.125,0.871948,0.349921,-74.012268,40.706718,N,0.000121
4,id2181028,2,2016-03-26 13:30:55,2016-03-26 13:38:10,0.125,0.872621,0.354186,-73.972923,40.782520,N,0.000123
5,id0801584,2,2016-01-30 22:01:40,2016-01-30 22:09:03,0.750,0.872443,0.351215,-73.992081,40.749184,N,0.000125
6,id1813257,1,2016-06-17 22:34:59,2016-06-17 22:40:40,0.500,0.872695,0.352126,-73.957405,40.765896,N,0.000096
7,id1324603,2,2016-05-21 07:54:58,2016-05-21 08:20:49,0.125,0.872690,0.354453,-73.922470,40.760559,N,0.000440
8,id1301050,1,2016-05-27 23:12:23,2016-05-27 23:16:38,0.125,0.872140,0.350994,-73.985786,40.732815,N,0.000072
9,id0012891,2,2016-03-10 21:45:01,2016-03-10 22:05:26,0.125,0.872476,0.351340,-73.973000,40.789989,N,0.000347


***Apply Standardization***

In [40]:
num_cols = [
    "passenger_count",
    "pickup_longitude",
    "pickup_latitude",
    "dropoff_longitude",
    "dropoff_latitude",
    "trip_duration"
]

In [41]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

df[num_cols] = scaler.fit_transform(df[num_cols])

In [43]:
df['pickup_longitude'].mean()

np.float64(1.3438355850833867e-14)

In [44]:
df[num_cols].agg(["mean", "std"])

,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,trip_duration
mean,-1.760770e-17,1.343836e-14,-7.187073e-15,-7.285129e-15,-5.956882e-16,1.126893e-17
std,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00


In [42]:
df.head()

,id,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration
0,id2875421,2,2016-03-14 17:24:55,2016-03-14 17:32:30,-0.505669,-0.124960,0.546173,0.130031,0.401659,N,-0.096486
1,id2377394,1,2016-06-12 00:43:35,2016-06-12 00:54:38,-0.505669,-0.099598,-0.398527,-0.380098,-0.603587,N,-0.056840
2,id3858529,2,2016-01-19 11:35:24,2016-01-19 12:10:48,-0.505669,-0.079353,0.417596,-0.465753,-1.218244,N,0.221631
3,id3504673,2,2016-04-06 19:32:31,2016-04-06 19:39:40,-0.505669,-0.531534,-0.996511,-0.567265,-1.316531,N,-0.101441
4,id2181028,2,2016-03-26 13:30:55,2016-03-26 13:38:10,-0.505669,0.007746,1.358984,0.008641,0.895322,N,-0.100298


***Compare both approaches***

<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #00bfff;
    border-radius: 12px;
    padding: 15px;
    margin: 5px 0;
    overflow: hidden;
    box-shadow:
        0 0 8px #00bfff,
        0 0 18px #0077ff,
        0 0 25px #ff1744;
    background: linear-gradient(
        135deg,
        rgba(0,191,255,0.06),
        rgba(255,23,68,0.06)
    );
">

<h1 style="
    text-align: center;
    color: #00bfff;
    font-size: 26px;
    text-shadow:
        0 0 8px #00bfff,
        0 0 15px #0077ff;
    margin: 5px 0 10px 0;
">
Compare Both Approaches
</h1>

<h3 style="
    text-align: center;
    color: #ff1744;
    font-size: 18px;
    text-shadow: 0 0 8px #ff1744;
    margin: 0 0 15px 0;
">
Min-Max Scaling vs Standardization
</h3>

<div style="
    width: 100%;
    overflow-x: auto;
">

<table style="
    width: 100%;
    table-layout: fixed;
    border-collapse: collapse;
    font-size: 13px;
">

<tr>
<th style="padding:8px; border:1px solid #00bfff; width:20%;">Point</th>
<th style="padding:8px; border:1px solid #00bfff; width:40%;">Min-Max Scaling</th>
<th style="padding:8px; border:1px solid #ff1744; width:40%;">Standardization</th>
</tr>

<tr>
<td style="padding:7px; border:1px solid #aaa;"><b>Purpose</b></td>
<td style="padding:7px; border:1px solid #aaa;">Values ko fixed range mein lana</td>
<td style="padding:7px; border:1px solid #aaa;">Values ko mean ke around standardize karna</td>
</tr>

<tr>
<td style="padding:7px; border:1px solid #aaa;"><b>Formula</b></td>
<td style="padding:7px; border:1px solid #aaa;">(X - min) / (max - min)</td>
<td style="padding:7px; border:1px solid #aaa;">(X - mean) / standard deviation</td>
</tr>

<tr>
<td style="padding:7px; border:1px solid #aaa;"><b>Output Range</b></td>
<td style="padding:7px; border:1px solid #aaa;"><b>Usually 0 to 1</b></td>
<td style="padding:7px; border:1px solid #aaa;">Fixed range nahi</td>
</tr>

<tr>
<td style="padding:7px; border:1px solid #aaa;"><b>Mean</b></td>
<td style="padding:7px; border:1px solid #aaa;">Zaroori nahi 0 ho</td>
<td style="padding:7px; border:1px solid #aaa;"><b>Approximately 0</b></td>
</tr>

<tr>
<td style="padding:7px; border:1px solid #aaa;"><b>Standard Deviation</b></td>
<td style="padding:7px; border:1px solid #aaa;">Zaroori nahi 1 ho</td>
<td style="padding:7px; border:1px solid #aaa;"><b>Approximately 1</b></td>
</tr>

<tr>
<td style="padding:7px; border:1px solid #aaa;"><b>Negative Values</b></td>
<td style="padding:7px; border:1px solid #aaa;">Normally nahi</td>
<td style="padding:7px; border:1px solid #aaa;"><b>Ho sakti hain</b></td>
</tr>

<tr>
<td style="padding:7px; border:1px solid #aaa;"><b>Outliers Effect</b></td>
<td style="padding:7px; border:1px solid #aaa;"><b>Zyada</b></td>
<td style="padding:7px; border:1px solid #aaa;">Relatively kam</td>
</tr>

<tr>
<td style="padding:7px; border:1px solid #aaa;"><b>Kab Useful?</b></td>
<td style="padding:7px; border:1px solid #aaa;">Jab fixed range chahiye</td>
<td style="padding:7px; border:1px solid #aaa;">Jab mean/std ke according scale karna ho</td>
</tr>

<tr>
<td style="padding:7px; border:1px solid #aaa;"><b>Example</b></td>
<td style="padding:7px; border:1px solid #aaa;">Neural Networks, Distance-based Methods</td>
<td style="padding:7px; border:1px solid #aaa;">Many ML Algorithms</td>
</tr>

</table>

</div>

</div>

<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **5: FEATURE ENCODING**

</div>

In [52]:
# First perform binning

bins=[0, 1, 2, 4, 6]
label=["Solo", "Small Group", "Medium Group", "Large Group"]
df['passenger_category'] = pd.cut(df['passenger_count'],bins=bins,labels=label)

***Use label encoding and one-hot encoding***

In [53]:
df = pd.get_dummies(
    df,
    columns=["passenger_category"],
    dtype=int
)


In [55]:
df.head(15)

,id,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration,passenger_category_Solo,passenger_category_Small Group,passenger_category_Medium Group,passenger_category_Large Group
0,id2875421,2,2016-03-14 17:24:55,2016-03-14 17:32:30,-0.505669,-0.124960,0.546173,0.130031,0.401659,N,-0.096486,0,0,0,0
1,id2377394,1,2016-06-12 00:43:35,2016-06-12 00:54:38,-0.505669,-0.099598,-0.398527,-0.380098,-0.603587,N,-0.056840,0,0,0,0
2,id3858529,2,2016-01-19 11:35:24,2016-01-19 12:10:48,-0.505669,-0.079353,0.417596,-0.465753,-1.218244,N,0.221631,0,0,0,0
3,id3504673,2,2016-04-06 19:32:31,2016-04-06 19:39:40,-0.505669,-0.531534,-0.996511,-0.567265,-1.316531,N,-0.101441,0,0,0,0
4,id2181028,2,2016-03-26 13:30:55,2016-03-26 13:38:10,-0.505669,0.007746,1.358984,0.008641,0.895322,N,-0.100298,0,0,0,0
5,id0801584,2,2016-01-30 22:01:40,2016-01-30 22:09:03,3.299634,-0.135194,-0.281728,-0.271774,-0.077422,N,-0.098773,0,0,1,0
6,id1813257,1,2016-06-17 22:34:59,2016-06-17 22:40:40,1.777513,0.066591,0.221417,0.235787,0.410230,N,-0.118215,0,1,0,0
7,id1324603,2,2016-05-21 07:54:58,2016-05-21 08:20:49,-0.505669,0.062809,1.505965,0.747145,0.254507,N,0.112415,0,0,0,0
8,id1301050,1,2016-05-27 23:12:23,2016-05-27 23:16:38,-0.505669,-0.377581,-0.403803,-0.179642,-0.555055,N,-0.134606,0,0,0,0
9,id0012891,2,2016-03-10 21:45:01,2016-03-10 22:05:26,-0.505669,-0.108831,-0.212777,0.007524,1.113268,N,0.050279,0,0,0,0


<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **6. NUMPY IMPLEMENTATION**

</div>

**Convert dataset into NumPy arrays**

In [56]:
data_array = df.to_numpy()
data_array = df.to_numpy()

**Check Type**

In [57]:
type(data_array)

numpy.ndarray

**Perform slicing and indexing***

In [60]:
data_array[0]

array(['id2875421', 2, Timestamp('2016-03-14 17:24:55'),
       Timestamp('2016-03-14 17:32:30'), -0.5056690804585793,
       -0.12496028431855817, 0.5461733952547639, 0.13003137973029832,
       0.40165926054268314, 'N', -0.09648579914603644, 0, 0, 0, 0],
      dtype=object)

In [61]:
data_array[0:5]

array([['id2875421', 2, Timestamp('2016-03-14 17:24:55'),
        Timestamp('2016-03-14 17:32:30'), -0.5056690804585793,
        -0.12496028431855817, 0.5461733952547639, 0.13003137973029832,
        0.40165926054268314, 'N', -0.09648579914603644, 0, 0, 0, 0],
       ['id2377394', 1, Timestamp('2016-06-12 00:43:35'),
        Timestamp('2016-06-12 00:54:38'), -0.5056690804585793,
        -0.09959809654528377, -0.39852742465203683, -0.3800981243356644,
        -0.6035867551654374, 'N', -0.05684036331324828, 0, 0, 0, 0],
       ['id3858529', 2, Timestamp('2016-01-19 11:35:24'),
        Timestamp('2016-01-19 12:10:48'), -0.5056690804585793,
        -0.07935284139285667, 0.41759593301293335, -0.465752530994726,
        -1.2182438804830389, 'N', 0.22163070280071095, 0, 0, 0, 0],
       ['id3504673', 2, Timestamp('2016-04-06 19:32:31'),
        Timestamp('2016-04-06 19:39:40'), -0.5056690804585793,
        -0.5315339523505614, -0.996510774811302, -0.5672647287171342,
        -1.31653113902256

 ***Apply aggregation functions***

In [65]:
df.columns.tolist()

['id',
 'vendor_id',
 'pickup_datetime',
 'dropoff_datetime',
 'passenger_count',
 'pickup_longitude',
 'pickup_latitude',
 'dropoff_longitude',
 'dropoff_latitude',
 'store_and_fwd_flag',
 'trip_duration',
 'passenger_category_Solo',
 'passenger_category_Small Group',
 'passenger_category_Medium Group',
 'passenger_category_Large Group']

In [66]:
data_array = df[['passenger_count', 'trip_duration']].to_numpy()

In [67]:
type(data_array)

numpy.ndarray

In [70]:
df[['passenger_count', 'trip_duration']].head()

,passenger_count,trip_duration
0,-0.505669,-0.096486
1,-0.505669,-0.056840
2,-0.505669,0.221631
3,-0.505669,-0.101441
4,-0.505669,-0.100298


In [71]:
df[['passenger_count', 'trip_duration']].mean()

passenger_count   -1.760770e-17
trip_duration      1.126893e-17
dtype: float64

In [ ]:
np.mean(data_array, axis=0)# Mean

array([-1.76077035e-17,  1.12689303e-17])

In [69]:
np.max(data_array, axis=0)

array([  4.82175557, 671.93691846])

In [73]:
df[['passenger_count', 'trip_duration']].min()

passenger_count   -1.26673
trip_duration     -0.18302
dtype: float64

In [72]:
np.min(data_array, axis=0)

array([-1.26672974, -0.18301959])

In [75]:
df[['passenger_count', 'trip_duration']].std()

passenger_count    1.0
trip_duration      1.0
dtype: float64

In [74]:
np.std(data_array, axis=0)

array([1., 1.])

***Perform Mathmatical Operation***

In [77]:
result = data_array[:, 0] + data_array[:, 1]
result

array([-0.60215488, -0.56250944, -0.28403838, ..., -0.54325854,
       -0.61778433, -0.65113987], shape=(1452747,))

In [79]:
result2 = data_array[:, 0] - data_array[:, 1]
result2

array([-0.40918328, -0.44882872, -0.72729978, ..., -0.46807963,
       -0.39355383, -0.3601983 ], shape=(1452747,))

In [81]:
result3 = data_array[:,0] * data_array[:,0]
result3

array([0.25570122, 0.25570122, 0.25570122, ..., 0.25570122, 0.25570122,
       0.25570122], shape=(1452747,))

<div style="
    width: 100%;
    box-sizing: border-box;
    border: 2px solid #2196f3;
    border-radius: 12px;
    padding: 15px 20px;
    margin: 10px 0;
    text-align: center;
    box-shadow: 0 0 8px #2196f3, 0 0 18px rgba(33,150,243,0.7);
">

# **7: ADVANCED VISUALIZATION (PLOTLY)**

</div>

 ***Create Geo map of pickup locations***

In [3]:
sample_df = df.sample(5000, random_state=42)

fig = px.scatter_map(
    sample_df,
    lat="pickup_latitude",
    lon="pickup_longitude",
    zoom=10,
    height=600,
    title="Taxi Pickup Locations"
)

fig.show()

In [4]:
import pandas as pd
import plotly.express as px

# Convert pickup_datetime to datetime
df['pickup_datetime'] = pd.to_datetime(df['pickup_datetime'])

# Extract date from pickup_datetime
df['pickup_date'] = df['pickup_datetime'].dt.date

# Count number of trips for each date
trip_counts = (
    df.groupby('pickup_date')
      .size()
      .reset_index(name='trips')
)

# Create animated plot
fig = px.bar(
    trip_counts,
    x='pickup_date',
    y='trips',
    animation_frame='pickup_date',
    title='Trips Over Time',
    labels={
        'pickup_date': 'Pickup Date',
        'trips': 'Number of Trips'
    }
)

fig.show()

In [5]:
import pandas as pd
import plotly.express as px

# Convert pickup_datetime to datetime
df['pickup_datetime'] = pd.to_datetime(df['pickup_datetime'])

# Extract date and hour
df['pickup_date'] = df['pickup_datetime'].dt.date
df['pickup_hour'] = df['pickup_datetime'].dt.hour

# Count trips by date and hour
trip_counts = (
    df.groupby(['pickup_date', 'pickup_hour'])
      .size()
      .reset_index(name='trips')
)

# Animated plot
fig = px.bar(
    trip_counts,
    x='pickup_hour',
    y='trips',
    animation_frame='pickup_date',
    range_x=[0, 23],
    range_y=[0, trip_counts['trips'].max()],
    title='Animated Taxi Trips Over Time',
    labels={
        'pickup_hour': 'Hour of Day',
        'trips': 'Number of Trips',
        'pickup_date': 'Date'
    }
)

fig.show()

In [6]:
import pandas as pd
import plotly.express as px

# Convert pickup_datetime to datetime
df['pickup_datetime'] = pd.to_datetime(df['pickup_datetime'])

# Passenger category
df['passenger_category'] = pd.cut(
    df['passenger_count'],
    bins=[0, 1, 2, 4, 6],
    labels=['Solo', 'Small Group', 'Medium Group', 'Large Group']
)

# Time category
df['hour'] = df['pickup_datetime'].dt.hour

df['time_category'] = pd.cut(
    df['hour'],
    bins=[-1, 5, 11, 17, 23],
    labels=['Night', 'Morning', 'Afternoon', 'Evening']
)

# Calculate trip distance using latitude and longitude
from math import radians, sin, cos, sqrt, atan2

def calculate_distance(row):
    lat1 = radians(row['pickup_latitude'])
    lon1 = radians(row['pickup_longitude'])
    lat2 = radians(row['dropoff_latitude'])
    lon2 = radians(row['dropoff_longitude'])

    dlat = lat2 - lat1
    dlon = lon2 - lon1

    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    c = 2 * atan2(sqrt(a), sqrt(1-a))

    return 6371 * c

df['distance_km'] = df.apply(calculate_distance, axis=1)

# Distance category
df['distance_category'] = pd.cut(
    df['distance_km'],
    bins=[0, 2, 5, 10, float('inf')],
    labels=['Short', 'Medium', 'Long', 'Very Long']
)

# Remove missing categories
sunburst_df = df.dropna(
    subset=['passenger_category', 'time_category', 'distance_category']
)

# Create Sunburst plot
fig = px.sunburst(
    sunburst_df,
    path=[
        'passenger_category',
        'time_category',
        'distance_category'
    ],
    title='Taxi Trips: Passenger → Time → Distance Category'
)

fig.show()

In [7]:
import pandas as pd
import plotly.express as px
import imageio.v2 as imageio
import os

# Convert pickup_datetime to datetime
df['pickup_datetime'] = pd.to_datetime(df['pickup_datetime'])

# Extract date and hour
df['pickup_date'] = df['pickup_datetime'].dt.date
df['pickup_hour'] = df['pickup_datetime'].dt.hour

# Count trips by date and hour
trip_counts = (
    df.groupby(['pickup_date', 'pickup_hour'])
      .size()
      .reset_index(name='trips')
)

# Create folder for frames
os.makedirs('gif_frames', exist_ok=True)

# Store frame file names
frames = []

# Create one image for each date
for i, date in enumerate(sorted(trip_counts['pickup_date'].unique())):

    data = trip_counts[trip_counts['pickup_date'] <= date]

    fig = px.bar(
        data,
        x='pickup_hour',
        y='trips',
        title=f'Taxi Trips Over Time - {date}',
        labels={
            'pickup_hour': 'Hour of Day',
            'trips': 'Number of Trips'
        }
    )

    fig.update_layout(
        xaxis=dict(dtick=1),
        yaxis=dict(range=[0, trip_counts['trips'].max() * 1.1])
    )

    frame_path = f'gif_frames/frame_{i:03d}.png'

    fig.write_image(frame_path)
    frames.append(frame_path)

# Create GIF
with imageio.get_writer(
    'taxi_trips_animation.gif',
    mode='I',
    duration=0.5
) as writer:

    for frame in frames:
        writer.append_data(imageio.imread(frame))

print("GIF saved successfully as taxi_trips_animation.gif")

ModuleNotFoundError: No module named 'imageio'

<style>
.conclusion-box {
    width: 92%;
    max-width: 1200px;
    margin: 25px auto;
    padding: 28px 32px;
    background: linear-gradient(145deg, #ffffff, #f8f9fa);
    border: 3px solid #FFD700;
    border-radius: 18px;
    position: relative;
    overflow: hidden;
    box-shadow: 0 0 15px rgba(255, 215, 0, 0.35);
    font-family: "Segoe UI", Arial, sans-serif;
    line-height: 1.7;
}

/* Animated yellow shine */
.conclusion-box::before {
    content: "";
    position: absolute;
    top: 0;
    left: -120%;
    width: 70%;
    height: 100%;
    background: linear-gradient(
        90deg,
        transparent,
        rgba(255, 215, 0, 0.25),
        rgba(255, 255, 255, 0.75),
        rgba(255, 215, 0, 0.25),
        transparent
    );
    transform: skewX(-25deg);
    animation: shine 3.5s infinite;
}

@keyframes shine {
    0% {
        left: -120%;
    }
    60% {
        left: 150%;
    }
    100% {
        left: 150%;
    }
}

.conclusion-box h1 {
    color: #222;
    text-align: center;
    font-size: clamp(28px, 4vw, 42px);
    margin-bottom: 25px;
    position: relative;
}

.conclusion-box h2 {
    color: #333;
    font-size: clamp(21px, 2.5vw, 29px);
    margin-top: 28px;
    border-left: 5px solid #FFD700;
    padding-left: 12px;
    position: relative;
}

.conclusion-box p {
    color: #444;
    font-size: clamp(15px, 1.4vw, 18px);
    position: relative;
}

.conclusion-box ul {
    position: relative;
    padding-left: 25px;
}

.conclusion-box li {
    color: #444;
    font-size: clamp(15px, 1.4vw, 18px);
    margin: 8px 0;
}

.conclusion-box strong {
    color: #222;
}

/* Responsive design */
@media (max-width: 768px) {
    .conclusion-box {
        width: 94%;
        padding: 20px;
    }
}
</style>

<div class="conclusion-box">

<h1>Conclusion</h1>

<h2>Project Summary</h2>

<p>
This project provided a practical experience of working with a real-world NYC Taxi dataset and applying different data science techniques from data preprocessing to advanced visualization.
</p>

<p>
During this project, I used <strong>Python and Pandas</strong> to load, explore, clean, and transform the dataset. I handled missing values and duplicate records, converted datetime columns, and removed invalid data where required. I also performed <strong>feature engineering</strong> by extracting useful information such as hour, day, and month from datetime data and calculating trip-related features, including distance using the <strong>Haversine formula</strong>.
</p>

<p>
For numerical data processing, I used <strong>NumPy</strong> to work with arrays, perform slicing and indexing, apply aggregation functions, and carry out mathematical operations. I also explored <strong>feature scaling</strong> by applying both <strong>Min-Max Scaling</strong> and <strong>Standardization</strong>, which helped me understand how different scaling techniques transform numerical features.
</p>

<p>
I worked with categorical data using <strong>Label Encoding</strong> and <strong>One-Hot Encoding</strong>, including the creation of meaningful passenger categories. This helped convert categorical information into a format that can be used effectively for further data analysis and machine learning tasks.
</p>

<p>
For visualization, I used <strong>Plotly</strong> to create interactive and advanced visualizations. These included a <strong>Geo Map of pickup locations</strong>, an <strong>animated visualization of trips over time</strong>, and a <strong>Sunburst chart</strong> showing the relationship between <strong>Passenger Category, Time, and Distance Category</strong>. I also saved a visualization as a <strong>GIF animation</strong> to present changes over time in a more understandable way.
</p>

<h2>What I Learned</h2>

<ul>
    <li>Data loading and exploration using <strong>Pandas</strong></li>
    <li>Data cleaning and handling missing values</li>
    <li>Removing duplicates and invalid records</li>
    <li>Datetime conversion and feature extraction</li>
    <li>Feature engineering and distance calculation</li>
    <li><strong>Haversine formula</strong> for geographical distance</li>
    <li><strong>Min-Max Scaling</strong> and <strong>Standardization</strong></li>
    <li>Label Encoding and One-Hot Encoding</li>
    <li>Numerical operations using <strong>NumPy</strong></li>
    <li>Interactive visualization using <strong>Plotly</strong></li>
    <li>Geo maps and animated visualizations</li>
    <li>Sunburst charts for hierarchical data</li>
    <li>Working with a real-world dataset from start to finish</li>
</ul>

<h2>Final Remarks</h2>

<p>
Overall, this project helped me understand how raw real-world data can be transformed into clean, structured, and meaningful information. It also improved my practical understanding of the complete data analysis workflow, from <strong>data loading and preprocessing to feature engineering, numerical analysis, and advanced visualization</strong>.
</p>

<p>
This project was a valuable step in developing my practical skills in <strong>Data Science, Data Analysis, Python, Pandas, NumPy, and Plotly</strong>.
</p>

</div>